all-sdk/external-client-configuration.md

```diff
@@ -0,0 +1,340 @@
+        * `serverName` - Override SNI name when connecting to server.
+        * `disableHostVerification` - Boolean to disable verifying TLS server host. May not be available in all Temporal
+          clients. 💭 Why? Not all SDKs offer this feature today.
+      * `codec` - Remote codec information to use for all encoding/decoding of payloads. May not be available in all
+        Temporal clients. 💭 Why? Not all SDKs allow use of a remote codec today.
+        * `endpoint` - Endpoint to the remote codec.
+        * `auth` - Authorization header for the remote codec.
+      * `grpcMeta` - Object representing HTTP headers. Values must be strings for now (can discuss arrays later).
+        ❓ Should we call this `headers`? We do in some cases and not others.
+  * 💭 Why JSON?
+    * Because YAML and TOML and others require a dependency in our SDKs. AWS for example has a homegrown INI-ish
+      implementation they have had to build in to every language.
+    * We accept that while JSON is technically human-authorable, it's not as easy as other formats. This is acceptable
+      since we will be providing tooling to mutate the config.
+    * ❓ Should we consider "json with comments" and some kind of very basic pre-parser in each language that strips
+      them? This could come later and not part of MVP.
+* This whole format will be defined in proto in the API repo in the `sdk` package.
+  * 💭 Why?
+    * Clear, centralized contract.
+    * Our SDKs already use proto dependency and this gives them stable JSON without requiring separate JSON dependency.
+    * Users can use the proto tooling to read/write configs programmatically since they are just proto JSON.
+* Do not validate that all keys are known in the file.
+  * 💭 Why?
+    * Originally this section was strict validation, but it became clear that we should allow newer-CLI-created config
+      files that may have newer keys to work with older SDKs.
+    * Similarly, it became clear that we're not validating environment variables this way, why config file keys?
+    * We want the tools that mutate this file to have strong validation on the keys being added, so hopefully this
+      avoids the concern of key typos.
+* All keys are optional in the file. Ideally in the final merged form, we'd always require `address` and we always
+  require `namespace` in namespace-specific clients (i.e. non cloud ops API), but it is unreasonable since we must fit
+  within SDK defaults.
+  * Originally this section did say `address` and `namespace` were required, but as SDK implementations came about, it
+    became clear that can't be done easily, so CLI/SDK defaults need to remain.
+* The default config file location is `~/.config/temporalio/temporal.json` for all platforms
+  * So in Windows this is often `C:\Users\MyUsername\.config\temporalio\temporal.json`
+  * ❓ Should we skip the subdir or make it `~/.config/temporal.json`? Unsure if there are any future files that would
```

Comment:
```suggestion
  * ❓ Should we skip the subdir and make it `~/.config/temporal.json`? Unsure if there are any future files that would
```

Reply:
Shouldn't this be `%APPDATA%` on windows? I think that's a bit more conventional.

Otherwise, yes I would definitely keep the `temporalio` subdir

Reply:
> Shouldn't this be %APPDATA% on windows? I think that's a bit more conventional.

Yes, I do think that is better and is what https://pkg.go.dev/os#UserConfigDir does even though CLI today used https://pkg.go.dev/os#UserHomeDir because it had to be compatible with `tctl`. I will change.

Reply:
Well, only Java and Core - see my other comment about having loading happen through Core. Pretty easy to do in Core.

Reply:
Updated